tests/test_helpers/dataset_utils.py

```diff
@@ -0,0 +1,258 @@
+    api_data: Dict[str, List[Row]],
+    file_name: str,
+):
+    """
+    Helper function to update the dataset in the given dataset_config
+    with api_data and write the formatted result to the specified file.
+    """
+
+    generated_dataset = generate_dataset(
+        dataset_config.dataset,
+        api_data,
+        [endpoint["name"] for endpoint in connection_config.saas_config["endpoints"]],
+    )
+
+    # the yaml library doesn't allow us to just reformat
+    # the data_categories field so we fix it with a regex
+    #
+    # data_categories:
+    #   - system.operations
+    #
+    # data_categories: [system.operations]
+    #
+    with open(f"{SAAS_DATASET_DIRECTORY}{file_name}", "w") as dataset_file:
+        dataset_file.write(
+            re.sub(
+                r"(data_categories:)\n\s+- ([^\n]+)",
+                r"\1 [\2]",
+                yaml.dump(
+                    {"dataset": [generated_dataset]},
+                    default_flow_style=False,
+                    sort_keys=False,
+                    indent=2,
+                ),
+            )
+        )
+
+
+def generate_dataset(
+    existing_dataset: Dict[str, Any],
+    api_data: Dict[str, List[Row]],
+    collection_order: List[str] = None,
+):
+    """
+    Generates a dataset which is an aggregate of the existing dataset and
+    any new fields generated from the API data. Orders the collections
+    based on the order of collection_order.
+    """
+
+    # preserve the collection order in the dataset if a collection order is not provided
+    if not collection_order:
+        collection_order = [
+            collection["name"] for collection in existing_dataset["collections"]
+        ]
+
+    # remove the dataset name from the keys in the api_data map before passing
+    # into generate_collections
+    generated_collections = generate_collections(
+        {
+            collection_name.replace(f"{existing_dataset['fides_key']}:", ""): collection
+            for collection_name, collection in api_data.items()
+        },
+        existing_dataset,
+    )
+
+    return {
+        "fides_key": existing_dataset["fides_key"],
+        "name": existing_dataset["name"],
+        "description": existing_dataset["description"],
+        "collections": [
+            {
+                "name": collection["name"],
+                "fields": collection["fields"],
+            }
+            for collection in sorted(
+                generated_collections,
+                key=lambda collection: collection_order.index(collection["name"]),
+            )
+        ],
+    }
+
+
+def generate_collections(
+    api_data: Dict[str, List[Row]], dataset: Dict[str, Any] = None
+) -> List[Dict[str, Any]]:
+    """
+    Generates a list of collections based on the response data or returns
+    the existing collections if no API data is available.
+    """
+
+    # convert FidesopsDataset to Dataset to be able to use the Collection helpers
```

Comment:
just noting that i don't totally follow this, but probably due to my lack of understanding, so may be good to review together offline